framework/include/auxkernels/MaterialAuxBase.h

```diff
@@ -50,26 +68,94 @@ InputParameters
 MaterialAuxBaseTempl<T, is_ad, RT>::validParams()
 {
   InputParameters params = AuxKernelTempl<RT>::validParams();
-  params.addRequiredParam<MaterialPropertyName>("property", "The scalar material property name");
+  params.addParam<MaterialPropertyName>(
+      "property", "The scalar material property name (set either this or `functor`).");
+  // functors do not support all types
+  if constexpr (!std::is_same_v<MaterialAuxFunctorType, void>)
+    params.addParam<MooseFunctorName>(
+        "functor", "The scalar material property name (set either this or `property`).");
   params.addParam<Real>(
       "factor", 1, "The factor by which to multiply your material property for visualization");
   params.addParam<RT>("offset", 0, "The offset to add to your material property for visualization");
+
+  params.addParam<unsigned int>(
+      "selected_qp",
+      "Evaluate the material property at a specified quadrature point. This only needs "
+      "to be used if you are interested in a particular quadrature point in each element. "
+      "Otherwise do not include this parameter in your input file.");
+  params.addParamNamesToGroup("selected_qp", "Advanced");
+
   return params;
 }
 
 template <typename T, bool is_ad, typename RT>
 MaterialAuxBaseTempl<T, is_ad, RT>::MaterialAuxBaseTempl(const InputParameters & parameters)
   : AuxKernelTempl<RT>(parameters),
-    _prop(this->template getGenericMaterialProperty<T, is_ad>("property")),
+    _property(this->isParamValid("property")
+                  ? &this->template getGenericMaterialProperty<T, is_ad>("property")
+                  : nullptr),
+    _functor(this->isParamValid("functor")
+                 ? [this]() {
+                    if constexpr (!std::is_same_v<MaterialAuxFunctorType, void>)
+                      return &this->template getFunctor<GenericType<T, is_ad>>("functor");
+                    else
+                    {
+                      libmesh_ignore(this);
+                      return nullptr;
+                    }
+                 }()
```

Comment:
This isn't necessary is it? The parameter won't be valid if it's not one of the supported types, so why not go ahead and always call `getFunctor` if the param is valid?

Reply:
Because that would instantiate the getFunctor template with an unsupported type! But I changed that design completely now.